server/src/lib/model/db.ts

```diff
@@ -586,4 +586,71 @@ export default class DB {
       );
     }
   }
+
+  async getUserPoints(client_id: string, challenge: string) {
+    const [[row]] = await this.mysql.query(
+      `
+      SELECT (bonus_points + clip_point + vote_point) AS user
+      FROM (
+          SELECT bonus_points, clip_point, COUNT(votes.id) AS vote_point
+          FROM (
+              SELECT challenger.client_id, start_date, end_date, bonus_points, COUNT(clips.id) AS clip_point
+              FROM (
+                  SELECT user_clients.client_id,
+                      challenges.start_date AS start_date,
+                      TIMESTAMPADD(WEEK, 3, challenges.start_date) AS end_date,
+                      SUM(achievements.points) AS bonus_points
+                  FROM user_clients
+                  LEFT JOIN enroll ON user_clients.client_id = enroll.client_id
+                  LEFT JOIN teams ON enroll.team_id = teams.id
+                  LEFT JOIN challenges ON teams.challenge_id = challenges.id AND enroll.challenge_id = challenges.id
+                  LEFT JOIN earn ON user_clients.client_id = earn.client_id
+                  LEFT JOIN achievements ON earn.achievement_id = achievements.id AND challenges.id = achievements.challenge_id
+                  WHERE user_clients.client_id = ?
+                  AND challenges.url_token = ?
+                  AND earn.earned_at BETWEEN challenges.start_date AND TIMESTAMPADD(WEEK, 3, challenges.start_date)
+                  GROUP BY user_clients.client_id, start_date, end_date
+              ) challenger
+              LEFT JOIN clips ON challenger.client_id = clips.client_id AND clips.created_at BETWEEN start_date AND end_date
+              GROUP BY challenger.client_id, start_date, end_date, bonus_points
+          ) speaker
+          LEFT JOIN votes ON speaker.client_id = votes.client_id AND votes.created_at BETWEEN start_date AND end_date
+          GROUP BY speaker.client_id, speaker.bonus_points, speaker.clip_point
+      ) voter
+      `,
+      [client_id, challenge]
+    );
+    return row;
+  }
+
+  async getWeeklyProgress(client_id: string, challenge: string) {
+    const [[row]] = await this.mysql.query(
+      `
+      SELECT speaker.client_id, start_date, end_date, week, colleague_count, clip_count, COUNT(votes.id) AS vote_count
+      FROM (
+          SELECT user.client_id, start_date, end_date, week, colleague_count, COUNT(clips.id) AS clip_count
+          FROM (
+              SELECT user_clients.client_id,
+                  TIMESTAMPDIFF(WEEK, start_date, NOW()) + 1 AS week,
+                  TIMESTAMPADD(WEEK, TIMESTAMPDIFF(WEEK, start_date, NOW()), start_date) AS start_date,
+                  TIMESTAMPADD(WEEK, TIMESTAMPDIFF(WEEK, start_date, NOW()) + 1, start_date) AS end_date,
+                  COUNT(colleagues.id) AS colleague_count
```

Comment:
Love it! If any of this ends up being too intensive for our DB at scale we can simplify and pass in a static `current_week` value, but this looks great for now.